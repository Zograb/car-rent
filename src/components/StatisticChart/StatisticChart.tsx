import type { FC } from 'react';
import type { DonutChartDataItem } from '@/types';
import 'twin.macro';

// Components
import { DonutChart } from '@/components/common/DonutChart';

interface StatisticCardProps {
  data: DonutChartDataItem[];
}

const StatisticChart: FC<StatisticCardProps> = ({ data }) => (
  <div tw="bg-white p-6 rounded-lg shadow-md">
    <div tw="flex justify-between items-center pb-3 text-gray3 mb-2">
      <div tw="font-medium text-lg">Hire vs Cancel</div>
      <div tw="bg-gray-100 py-1 px-2 rounded-sm text-xs">Today</div>
    </div>
    <DonutChart data={data} showLegend />
  </div>
);

export default StatisticChart;
