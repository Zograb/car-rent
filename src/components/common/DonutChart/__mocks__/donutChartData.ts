const donutChartData = [
  {
    name: 'Part 1',
    value: 360,
    color: '#52C93F',
  },
  {
    name: 'Part 2',
    value: 230,
    color: '#006AFF',
  },
  {
    name: 'Part 3',
    value: 150,
    color: '#FF2727',
  },
  {
    name: 'Part 4',
    value: 230,
    color: '#FF2727',
  },
  {
    name: 'Part 5',
    value: 150,
    color: '#FFFF00',
  },
];

export default donutChartData;
