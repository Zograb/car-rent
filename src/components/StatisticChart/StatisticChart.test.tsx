import type { ReactElement } from 'react';
import { render, screen } from '@testing-library/react';
import recharts from 'recharts';

// Components
import StatisticChart from './StatisticChart';

// Mocks
import donutChartData from '@/components/common/DonutChart/__mocks__/donutChartData';
import {
  mockResizeObserverAfter,
  mockResizeObserverBefore,
} from '@/__mocks__/';

// Set responsive container width and height to fixed value
// Otherwise in test envirinment it will not render chart at all
jest.mock('recharts', () => {
  const OriginalModule: typeof recharts = jest.requireActual('recharts');

  return {
    ...OriginalModule,
    ResponsiveContainer: ({ children }: { height: number, children: ReactElement }) => (
      <div style={{ width: 500, height: 500 }}>
        {children}
      </div>
    ),
  };
});

describe('StatisticChart', () => {
  beforeEach(() => {
    mockResizeObserverBefore();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <StatisticChart data={donutChartData} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render statistic chart', () => {
    const { container } = render(
      <StatisticChart data={donutChartData} />
    );
    const statisticChartTitle = screen.getByText('Hire vs Cancel');
    const chart = container.querySelector('svg');
    expect(statisticChartTitle).toBeInTheDocument();
    expect(chart).toBeInTheDocument();
  });

  afterEach(mockResizeObserverAfter);
});
