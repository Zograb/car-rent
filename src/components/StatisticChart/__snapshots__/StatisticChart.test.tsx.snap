// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticChart should match snapshot 1`] = `
.c4 {
  height: 11rem;
  width: 11rem;
  margin-left: auto;
  margin-right: auto;
}

.c5 {
  margin-top: 1rem;
}

.c6 {
  margin-bottom: 0.5rem;
  display: flex;
  align-items: center;
}

.c7 {
  height: 1rem;
  width: 1rem;
  border-radius: 9999px;
}

.c8 {
  margin-left: 0.75rem;
  flex-grow: 1;
  text-transform: capitalize;
}

.c0 {
  border-radius: 0.5rem;
  --tw-bg-opacity: 1;
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
  padding: 1.5rem;
  --tw-shadow: 0 4px 6px -1px rgb(0 0 0 / 0.1),0 2px 4px -2px rgb(0 0 0 / 0.1);
  --tw-shadow-colored: 0 4px 6px -1px var(--tw-shadow-color),0 2px 4px -2px var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
}

.c1 {
  margin-bottom: 0.5rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding-bottom: 0.75rem;
  --tw-text-opacity: 1;
  color: rgb(101 101 117 / var(--tw-text-opacity));
}

.c2 {
  font-size: 1.125rem;
  line-height: 1.75rem;
  font-weight: 500;
}

.c3 {
  border-radius: 0.125rem;
  --tw-bg-opacity: 1;
  background-color: rgb(243 244 246 / var(--tw-bg-opacity));
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  font-size: 0.75rem;
  line-height: 1rem;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Hire vs Cancel
      </div>
      <div
        class="c3"
      >
        Today
      </div>
    </div>
    <div
      class=""
    >
      <div
        class="c4"
      >
        <div
          style="width: 500px; height: 500px;"
        >
          <div
            class="recharts-wrapper"
            role="region"
            style="position: relative; cursor: default; width: 400px; height: 400px;"
          >
            <svg
              class="recharts-surface"
              cx="50%"
              cy="50%"
              height="400"
              viewBox="0 0 400 400"
              width="400"
            >
              <title />
              <desc />
              <defs>
                <clippath
                  id="recharts1-clip"
                >
                  <rect
                    height="390"
                    width="390"
                    x="5"
                    y="5"
                  />
                </clippath>
              </defs>
              <g
                class="recharts-layer recharts-pie"
                tabindex="0"
              >
                <g
                  class="recharts-layer"
                >
                  <g
                    class="recharts-layer recharts-pie-sector"
                    tabindex="-1"
                  />
                  <g
                    class="recharts-layer recharts-pie-sector"
                    tabindex="-1"
                  />
                  <g
                    class="recharts-layer recharts-pie-sector"
                    tabindex="-1"
                  />
                  <g
                    class="recharts-layer recharts-pie-sector"
                    tabindex="-1"
                  />
                  <g
                    class="recharts-layer recharts-pie-sector"
                    tabindex="-1"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        class="c5"
        data-testid="donut-chart-legend"
      >
        <div
          class="c6"
        >
          <div
            class="c7"
            style="background-color: rgb(82, 201, 63);"
          />
          <div
            class="c8"
          >
            Part 1
          </div>
          <div>
            32%
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
            style="background-color: rgb(0, 106, 255);"
          />
          <div
            class="c8"
          >
            Part 2
          </div>
          <div>
            21%
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
            style="background-color: rgb(255, 39, 39);"
          />
          <div
            class="c8"
          >
            Part 3
          </div>
          <div>
            13%
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
            style="background-color: rgb(255, 39, 39);"
          />
          <div
            class="c8"
          >
            Part 4
          </div>
          <div>
            21%
          </div>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
            style="background-color: rgb(255, 255, 0);"
          />
          <div
            class="c8"
          >
            Part 5
          </div>
          <div>
            13%
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
